### PR TITLE
CMake: Export targets so the build directory can be used directly

### PR DESCRIPTION
**Pull request by ZedThree**
Along with #2751, this fully enables using the build directory directly in other projects, allowing other projects to include netCDF using things like `FetchContent`.

I've had to delete the paths in `netCDFConfig.cmake.in` because they won't exist until netCDF is installed (sort of defeating the point of using the build directory!) -- this should be fine as long as people are using the `netCDF::netcdf` target rather than these variables. It might be possible to support both, but it might get a bit complicated!

### CODE REVIEW

**Comment by WardF** [comment]
We historically discourage people from using the build directory directly, and encourage people to use the install directory instead; this has been a best practice for a very long time.  I'm not opposed to making it easier for people to use their `build/` directory if that is what they want to do, but I'm not certain what you mean, specifically, by `It might be possible to support both, but it might get a bit complicated`.  Forgive my obtuseness, I'd like to merge this in, I just want to make sure I understand what you mean.  I presume this PR does not make it more difficult for people who use the standard `make && make install` process?

**Comment by ZedThree** [comment]
The major reason to enable using the build directory directly is to allow embedding in another project, for example as part of a superbuild or a bundled dependency. There are at least a couple of ways to do this in practice, either with a git submodule (or just bundled) and `add_subdirectory`, or using `FetchContent` to download it as part of the build process. There are also various CMake package managers like [CPM](https://github.com/cpm-cmake/CPM.cmake) and [vcpkg](https://vcpkg.io/en/), which generally work using `FetchContent` -- some of these tools already patch netCDF to make this work.

Basically, in modern CMake, it's a really nice user experience to be able to use the build directory for a project, as well as installing it.

> It might be possible to support both, but it might get a bit complicated

Sorry, I was not very clear here! Just so we're all on the same page (and for others reading this), pre-CMake 3 and the introduction of targets, the standard way of using a dependency was with variables like `package_LIBRARIES` and `package_INCLUDE_DIRS`. Modern CMake wraps these all up into a target like `package::package`. The current `netCDFConfig.cmake` file that gets installed sets **both** the target and the `netCDF_LIBRARIES/INCLUDE_DIR` variables (as well as `netCDF_INSTALL_PREFIX` and `netCDF_LIB_DIR`) -- but only the target are required.

So, if we delete the `netCDF_*` variables from the config file, this might possibly break people's CMake files -- but the fix is to switch to use the targets instead, so pretty easy. The targets have been available since 2014 in netCDF v4.3.3, so quite some time!

It is possible to keep the variables, but it would involve generating the config file twice, once for the build directory, once for the install directory. Not a problem, it's just more code to do so!

> I presume this PR does not make it more difficult for people who use the standard `make && make install` process? 

Nope! Just adds another way to consume netCDF

**Comment by LecrisUT** [comment]
I don't think that this will address using `FetchContent` because that would require running `find_package` after `FetchContent` in order to use `netCDFConfig.cmake` from the build directory which will clash with `FetchContent(FIND_PACKAGE_ARGS)`.

Regardless, this change is still necessary to do functional tests of examples, e.g. to test the packaging [example](https://github.com/LecrisUT/CMake-Template/blob/a4f73dbd53d8cb98e6325296c2905b2ed40dcaa3/test/CMakeLists.txt#L114-L131).

If you want to address `FetchContent`, some necessary points are:
- Make sure relevant `target_include_directories`, etc. have appropriate `PUBLIC` interface where needed
- Make sure the options and targets are namespaces, with the latter having `add_library(ALIAS)`
- Make sure you export the `netCDF_VERSION` and other such variables that are available in `netCDFConfig.cmake` via `return(PROPAGATE)`

File: CMakeLists.txt [hunk 1]
```diff
@@ -2745,6 +2745,11 @@ install(EXPORT netCDFTargets
   NAMESPACE netCDF::
   )
 
+export(EXPORT netCDFTargets
+  FILE "${CMAKE_CURRENT_BINARY_DIR}/netCDFTargets.cmake"
```
Comment (LecrisUT) [inline CMakeLists.txt:2749]: `${CMAKE_CURRENT_BINARY_DIR}` is unnecessary if you want to simplify it like you did with `write_basic_package_version_file`

File: CMakeLists.txt [hunk 2]
```diff
@@ -2772,14 +2777,14 @@ target_include_directories(netcdf
...
-  "${CMAKE_CURRENT_BINARY_DIR}/netCDF/netCDFConfigVersion.cmake"
+  netCDFConfigVersion.cmake
   VERSION ${netCDF_VERSION}
   COMPATIBILITY SameMajorVersion
   )
 
 install(
   FILES
-  "${CMAKE_CURRENT_BINARY_DIR}/netCDF/netCDFConfigVersion.cmake"
+  "${CMAKE_CURRENT_BINARY_DIR}/netCDFConfigVersion.cmake"
```
Comment (LecrisUT) [inline CMakeLists.txt:2787]: Can you move `"${CMAKE_CURRENT_BINARY_DIR}/netCDFConfig.cmake"` here as well?

**Review by LecrisUT — APPROVED** [review]
There is also `configure_package_config_file` which seems excessive with
```
  PATH_VARS
  CMAKE_INSTALL_PREFIX
  CMAKE_INSTALL_INCLUDEDIR
  CMAKE_INSTALL_LIBDIR
```
And since it uses `PACKAGE_INIT` it should no have `NO_CHECK_REQUIRED_COMPONENTS_MACRO`

Otherwise, functionally :+1: this PR should work as is

**Comment by LecrisUT** [comment]
Just some short clarifications:

> which as you note is a necessary but not sufficient condition for FetchContent to also work.

FetchContent has different inclusion methods, i.e. it uses `add_subdirectory` instead of `find_package(CONFIG)`, so getting the project to work for either is an independent task

> check several ways of consuming the package (build dir vs install, for example)

All except for install. For that one I run tests on Fedora packaging on top of that (which calls the same test-suite). See the `tmt` + `packit` setup if you're curious on how to set it up. Probably I need more explicit documentation on that though.

> `return(PROPAGATE)` would require bumping the min cmake version

Not really, the [workaround](https://github.com/LecrisUT/CMake-Template/blob/a4f73dbd53d8cb98e6325296c2905b2ed40dcaa3/CMakeLists.txt#L169-L184) is a bit verbose but this one you can support both versions with policy `CMP0140`.

**Comment by K20shores** [comment]
@ZedThree yes, I am looking into those things you mentioned

**Pull request merged**